chain-signatures/contract-eth/contracts/ChainSignatures.sol

```diff
@@ -27,144 +25,55 @@ contract ChainSignatures is Initializable {
+        bytes32 requestId = keccak256(abi.encodePacked(payload, msg.sender, path, keyVersion, algo, dest, params));
 
-        uint256 epsilon = deriveEpsilon(path, msg.sender);
-        PublicKey memory derivedPk = _request.derivedPublicKey;
-        require(verifyDerivedKey(epsilon, derivedPk), "Derived key verification failed");
-        SignatureRequest memory request = SignatureRequest(epsilon, payloadHash, msg.sender, derivedPk);
-        pendingRequests[requestId] = request;
-        depositToRefund[requestId] = msg.value - requiredDeposit;
-        requestCounter++;
-
-        emit SignatureRequested(requestId, msg.sender, epsilon, payloadHash, path);
+        emit SignatureRequested(requestId, msg.sender, payload, keyVersion, msg.value, path, algo, dest, params);
 
         return requestId;
     }
     
-    function respond(bytes32 _requestId, SignatureResponse memory _response) external {        
-        SignatureRequest storage request = pendingRequests[_requestId];
-        require(request.requester != address(0), "Request not found");
-
-        PublicKey memory expectedPublicKey = request.derivedPublicKey;
-        // Derive Ethereum address from public key
-        bytes32 pkHash = keccak256(abi.encodePacked(expectedPublicKey.x, expectedPublicKey.y));
-        address expectedSigner = address(uint160(uint256(pkHash)));
-        // Verify the virtual signer is the derived address using ecrecover
-        address recoveredSigner = ecrecover(bytes32(request.payloadHash), _response.recoveryId+27, bytes32(_response.bigR.x), bytes32(_response.s));
-        require(recoveredSigner == expectedSigner, "Invalid signature");
-
-        emit SignatureResponded(_requestId, _response);
-
-        // Refund excess deposit
-        uint256 refund = depositToRefund[_requestId];
+    function respond(ResponseWithId[] calldata _responses) external {
+        for (uint256 i = 0; i < _responses.length; i++) {
+            emit SignatureResponded(_responses[i].requestId, _responses[i].response);
+        }
+    }
 
-        // Clean up
-        delete pendingRequests[_requestId];
-        delete depositToRefund[_requestId];
-        requestCounter--;
+    function getSignatureDeposit() external view returns (uint256) {
+        return signatureDeposit;
+    }
 
-        if (refund > 0) {
-            payable(request.requester).transfer(refund);
-        }
+    function setSignatureDeposit(uint256 _amount) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        signatureDeposit = _amount;
     }
 
-    function getSignatureDeposit() public view returns (uint256) {
-        // Simplified deposit calculation
-        if (requestCounter <= 3) {
-            return 1 wei;
-        } else {
-            return (requestCounter - 3) * 4 * 1e15; // 0.004 ETH (~1 USD) first request after the first 3
-        }
+    function withdraw(uint256 _amount, address _receiver) external onlyRole(DEFAULT_ADMIN_ROLE) {
```

Comment:
do you think we might want to do a emit withdrawal event here? Not sure what's good practice but from what ive seen, it might be better for things like etherscan

Reply:
added